metaphor/openapi/extractor.py

```diff
@@ -0,0 +1,159 @@
+
+import requests
+
+from metaphor.common.base_extractor import BaseExtractor
+from metaphor.common.event_util import ENTITY_TYPES
+from metaphor.common.logger import get_logger
+from metaphor.common.utils import md5_digest
+from metaphor.models.crawler_run_metadata import Platform
+from metaphor.models.metadata_change_event import (
+    API,
+    APILogicalID,
+    APIPlatform,
+    AssetPlatform,
+    AssetStructure,
+    Hierarchy,
+    HierarchyInfo,
+    HierarchyLogicalID,
+    HierarchyType,
+    OpenAPI,
+    OpenAPIMethod,
+    OpenAPISpecification,
+    OperationType,
+)
+from metaphor.openapi.config import OpenAPIRunConfig
+
+logger = get_logger()
+
+
+class OpenAPIExtractor(BaseExtractor):
+    """OpenAPI metadata extractor"""
+
+    _description = "OpenAPI metadata crawler"
+    _platform = Platform.OPEN_API
+
+    @staticmethod
+    def from_config_file(config_file: str) -> "OpenAPIExtractor":
+        return OpenAPIExtractor(OpenAPIRunConfig.from_yaml_file(config_file))
+
+    def __init__(self, config: OpenAPIRunConfig):
+        super().__init__(config)
+
+        self._base_url = config.base_url
+        self._api_id = md5_digest(config.base_url.encode("utf-8"))
+        self._openapi_json_path = config.openapi_json_path
+        self._auth = config.auth
+        self._requests_session = requests.sessions.Session()
+
+    async def extract(self) -> Collection[ENTITY_TYPES]:
+        logger.info(f"Fetching metadata from {self._openapi_json_path}")
+
+        self._initial_session()
+        openapi_json = self._get_openapi_json()
+
+        if not openapi_json:
+            logger.error("Unable to get OAS json")
+            return []
+
+        endpoints = self._extract_paths(openapi_json)
+        hierarchies = self._build_hierarchies(openapi_json)
+
+        return hierarchies + endpoints
+
+    def _initial_session(self):
+        if not self._auth:
+            return
+
+        if self._auth.basic_auth:
+            basic_auth = self._auth.basic_auth
+            self._requests_session.auth = (basic_auth.user, basic_auth.password)
+
+    def _get_openapi_json(self) -> Optional[dict]:
+        if not self._openapi_json_path.startswith("http"):
+            with open(self._openapi_json_path, "r") as f:
+                return json.load(f)
+
+        headers = OrderedDict(
+            {
+                "User-Agent": None,
+                "Accept": None,
+                "Connection": None,
+                "Accept-Encoding": None,
+            }
+        )
+        resp = self._requests_session.get(self._openapi_json_path, headers=headers)
+
+        if resp.status_code != 200:
+            return None
+
+        return resp.json()
+
+    def _extract_paths(self, openapi: dict) -> List[API]:
+        endpoints: List[API] = []
+        servers = openapi.get("servers")
+
+        for path, path_item in openapi["paths"].items():
+            path_servers = path_item.get("servers")
+            base_path = (
```

Comment:
```suggestion
            server_url_path = (
```

joining `self._base_url` with `base_path` is a little confusing